packages/core/src/agents/executor.ts

```diff
@@ -0,0 +1,578 @@
+        const text =
+          parts
+            ?.filter((p) => !p.thought && p.text)
+            .map((p) => p.text)
+            .join('') || '';
+
+        if (text) {
+          textResponse += text;
+        }
+      }
+    }
+
+    return { functionCalls, textResponse };
+  }
+
+  /** Initializes a `GeminiChat` instance for the agent run. */
+  private async createChatObject(inputs: AgentInputs): Promise<GeminiChat> {
+    const { promptConfig, modelConfig } = this.definition;
+
+    if (!promptConfig.systemPrompt && !promptConfig.initialMessages) {
+      throw new Error(
+        'PromptConfig must define either `systemPrompt` or `initialMessages`.',
+      );
+    }
+
+    const startHistory = [...(promptConfig.initialMessages ?? [])];
+
+    // TODO: Evaluate if subagents need environment context (date, OS, folder structure)
+    // Currently commented out to reduce token usage - subagents should focus on specific tasks
+    // const envParts = await getEnvironmentContext(this.runtimeContext);
+    // const envHistory: Content[] = [
+    //   { role: 'user', parts: envParts },
+    //   { role: 'model', parts: [{ text: 'Got it. Thanks for the context!' }] },
+    // ];
+
+    // Build system instruction from the templated prompt string.
+    const systemInstruction = promptConfig.systemPrompt
+      ? this.buildSystemPrompt(inputs)
+      : undefined;
+
+    try {
+      const generationConfig: GenerateContentConfig = {
+        temperature: modelConfig.temp,
+        topP: modelConfig.top_p,
+        thinkingConfig: {
+          includeThoughts: true,
+          thinkingBudget: modelConfig.thinkingBudget ?? -1,
+        },
+      };
+
+      if (systemInstruction) {
+        generationConfig.systemInstruction = systemInstruction;
+      }
+
+      return new GeminiChat(
+        this.runtimeContext,
+        generationConfig,
+        startHistory,
+      );
+    } catch (error) {
+      await reportError(
+        error,
+        `Error initializing Gemini chat for agent ${this.definition.name}.`,
+        startHistory,
+        'startChat',
+      );
+      // Re-throw as a more specific error after reporting.
+      throw new Error(`Failed to create chat object: ${error}`);
+    }
+  }
+
+  /**
+   * Executes function calls requested by the model and returns the results.
+   *
+   * @returns A new `Content` object to be added to the chat history.
+   */
+  private async processFunctionCalls(
+    functionCalls: FunctionCall[],
+    signal: AbortSignal,
+    promptId: string,
+  ): Promise<Content[]> {
+    const allowedToolNames = new Set(this.toolRegistry.getAllToolNames());
+
+    // Filter out any tool calls that are not in the agent's allowed list.
+    const validatedFunctionCalls = functionCalls.filter((call) => {
+      if (!allowedToolNames.has(call.name as string)) {
+        console.warn(
+          `[AgentExecutor] Agent '${this.definition.name}' attempted to call ` +
```

Comment:
why is this not treated like any other function call in the parent agent? i.e if a tool call is not present in the registered list than it is a hallucinated tool call.